fairseq/training.py

```diff
@@ -0,0 +1,491 @@
+        :param FairseqHydraConfig other:
+        """
+        self.config_dict = util.update_nested_dict(self.config_dict, other.config_dict)
+        self.post_config_dict = util.update_nested_dict(
+            self.post_config_dict, other.post_config_dict
+        )
+        if other.package_name != "":
+            self.package_name = other.package_name
+        self.check_consistency()
+
+    def check_consistency(self):
+        """
+        Check that there is no config key overwritten by post_config.
+        Also check for parameters that should never be hashed.
+        """
+        for group in self.config_dict:
+            if isinstance(self.config_dict[group], dict):
+                for key in self.config_dict[group]:
+                    if group in self.post_config_dict:
+                        assert (
+                            key not in self.post_config_dict[group].keys()
+                        ), f"{key} of {group} in post_config would overwrite existing entry in config"
+            else:
+                assert (
+                    group not in self.post_config_dict
+                ), f"{group} in post_config would overwrite existing entry in config"
+
+        # list of parameters that should never be hashed
+        disallowed_in_config = ["save_interval", "max_epoch"]
+
+        for group in self.config_dict:
+            if isinstance(self.config_dict[group], dict):
+                for key in disallowed_in_config:
+                    assert (
+                        self.config_dict[group].get(key) is None
+                    ), f"please define {key} of {group} only as parameter in the post_config_dict"
+            else:
+                assert (
+                    self.config_dict[group].get(group) is None
+                ), f"please define {group} only as parameter in the post_config_dict"
+
+    def _sis_hash(self):
+        h = {"fairseq_hydra_config": self.config_dict}
+        return sis_hash_helper(h)
+
+
+class PytorchHydraModel:
+    """
+    Defines a Pytorch hydra model as yaml config, pytorch checkpoint file and epoch
+    """
+
+    def __init__(self, fairseq_hydra_config_file: tk.Path, model: tk.Path, epoch: int):
+        """
+        :param fairseq_hydra_config_file: Path to a returnn config file
+        :param model: Path to a pytorch checkpoint
+        :param epoch: Number of epochs this model was trained
+        """
+        self.returnn_config_file = fairseq_hydra_config_file
+        self.model = model
+        self.epoch = epoch
+
+
+class FairseqHydraTrainingJob(Job):
+    """
+    Train a Fairseq model using fairseq-hydra-train
+    """
+
+    def __init__(
+        self,
+        fairseq_hydra_config,
+        *,  # args below are keyword only
+        command_line_args=None,
+        max_epoch=1,
+        save_interval=1,
+        keep_epochs=None,
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=2,
+        gpu_rqmt=1,
```

Comment:
Do we maybe also want to allow setting `gpu_mem_rqmt`? This would be helpful for our slurm engine.